compiler_opt/rl/env.py

```diff
@@ -0,0 +1,373 @@
+
+    def _get_step_type():
+      step_type = FIRST_STEP_STR if self._is_first_obs else MID_STEP_STR
+      self._is_first_obs = False
+      return step_type
+
+    try:
+      obs: log_reader.ObservationRecord = next(self._obs_gen)
+
+      tv_dict = {}
+      for fv in obs.feature_values:
+        array = fv.to_numpy()
+        tv_dict[fv.spec.name] = np.reshape(array, newshape=fv.spec.shape)
+      return {
+          OBS_KEY: tv_dict,
+          REWARD_KEY: obs.score if obs.score else 0.0,
+          SCORE_POLICY_KEY: 0.0,
+          SCORE_DEFAULT_KEY: 0.0,
+          CONTEXT_KEY: obs.context,
+          MODULE_NAME_KEY: self._module_name,
+          OBS_ID_KEY: obs.observation_id,
+          STEP_TYPE_KEY: _get_step_type(),
+      }
+    except StopIteration:
+      return self._terminal_obs
+
+  def send_action(self, action: np.ndarray) -> None:
+    assert self._running()
+    data = action.tobytes()
+    bytes_sent = self._writer_pipe.write(data)
+    # Here we use the fact that for common types, the np.dtype and ctype should
+    # behave the same
+    assert bytes_sent == action.dtype.itemsize * math.prod(action.shape)
+    try:
+      self._writer_pipe.flush()
+    except BrokenPipeError:
+      # The pipe can break after we send the last action
+      pass
+
+
+_EPS = 1e-4
+
+
+def compute_relative_rewards(score_a: dict[str, float],
+                             score_b: dict[str, float]) -> dict[str, float]:
+
+  def _reward_fn(a: float, b: float) -> float:
+    return 1.0 - (a + _EPS) / (b + _EPS)
+
+  assert score_a.keys() == score_b.keys()
+  return {key: _reward_fn(score_a[key], score_b[key]) for key in score_a}
+
+
+@contextlib.contextmanager
+def clang_session(
+    clang_path: str,
+    module: corpus.LoadedModuleSpec,
+    task_factory: Callable[[str], MLGOTask],
+    *,
+    interactive: bool,
+):
+  """Context manager for clang session.
+
+  We need to manage the context so resources like tempfiles and pipes have
+  their lifetimes managed appropriately.
+
+  Args:
+    clang_path: The clang binary to use for the InteractiveClang session.
+    module: The module to compile with clang.
+    task_factory: Callable which produces MLGOTask objects,
+    interactive: Whether to use an interactive or default clang instance
+
+  Yields:
+    Either the constructed InteractiveClang or DefaultClang object.
+  """
+  with tempfile.TemporaryDirectory() as td:
+    task_working_dir = os.path.join(td, '__task_working_dir__')
+    os.mkdir(task_working_dir)
+    task = task_factory(task_working_dir)
+
+    compiled_module_path = os.path.join(td, _COMPILED_MODULE_NAME)
+    cmdline = ([clang_path] + list(module.build_command_line(td)) +
+               task.get_clang_args() + ['-o', compiled_module_path])
+    if interactive:
+      cmdline += task.get_interactive_clang_args(
+          os.path.join(td, _INTERACTIVE_PIPE_FILE_BASE))
+
+    def _get_scores() -> dict[str, float]:
+      assert os.path.exists(compiled_module_path)
```

Comment:
I'd let the MLGOTask figure out where to get the scores from.

Reply:
they can get the scores from wherever they want, this is just making sure that we actually finishing compiling the module before calling the method